bindings/c/src/lib.rs

```diff
@@ -0,0 +1,202 @@
+    let ptr: *mut c_char = buffer.buffer.add(buffer.pos);
+    *ptr = 0;
+    CssResult::Ok
+}
+
+/// @brief Creates an instance of CssInlinerOptions with the default parameters.
+/// @return a CssInlinerOptions struct
+#[no_mangle]
+pub extern "C" fn css_inliner_default_options() -> CssInlinerOptions {
+    CssInlinerOptions {
+        keep_style_tags: false,
+        keep_link_tags: false,
+        base_url: null(),
+        load_remote_stylesheets: true,
+        extra_css: null(),
+        preallocate_node_capacity: 32,
+    }
+}
+
+struct CBuffer {
+    buffer: *mut c_char,
+    size: size_t,
+    pos: usize,
+}
+
+impl TryFrom<&CssInlinerOptions> for InlineOptions<'_> {
+    type Error = InlineOptionsError;
+
+    fn try_from(value: &CssInlinerOptions) -> Result<Self, Self::Error> {
+        let base_url: Option<&str> = unsafe {
+            // .as_ref() returns None when the pointer is null
+            match value.base_url.as_ref() {
+                Some(val) => Some(match CStr::from_ptr(val).to_str() {
+                    Ok(val) => val,
+                    Err(_) => return Err(InlineOptionsError::InvalidUrl),
+                }),
+                None => None,
+            }
+        };
+        let extra_css: Option<&str> = unsafe {
+            // .as_ref() returns None when the pointer is null
+            match value.extra_css.as_ref() {
+                Some(val) => Some(match CStr::from_ptr(val).to_str() {
+                    Ok(val) => val,
+                    Err(_) => return Err(InlineOptionsError::InvalidExtraCss),
+                }),
+                None => None,
+            }
+        };
+        Ok(Self {
+            keep_style_tags: value.keep_style_tags,
+            keep_link_tags: value.keep_link_tags,
+            base_url: match base_url {
+                Some(url) => Some(Url::parse(url).map_err(|_| InlineOptionsError::InvalidUrl)?),
+                None => None,
+            },
+            load_remote_stylesheets: value.load_remote_stylesheets,
+            extra_css: extra_css.map(Cow::Borrowed),
+            preallocate_node_capacity: value.preallocate_node_capacity,
+        })
+    }
+}
+
+impl From<InlineOptionsError> for CssResult {
+    fn from(value: InlineOptionsError) -> Self {
+        match value {
+            InlineOptionsError::InvalidUrl => CssResult::InvalidUrl,
+            InlineOptionsError::InvalidExtraCss => CssResult::InvalidExtraCss,
+        }
+    }
+}
+
+impl CBuffer {
+    fn new(buffer: *mut c_char, size: size_t) -> Self {
+        Self {
+            buffer,
+            size,
+            pos: 0,
+        }
+    }
+}
+
+impl Write for CBuffer {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        // Because write() attempts to write regardless of buf.len() being higher than self.size,
+        // we are going to check at each byte that we didn't cross the bounds
+        let mut bytes_written: usize = 0;
+        for byte in buf {
```

Comment:
I am thinking about whether we can think about writing in chunks and if it would be too much of a complication and how much would it affect performance

Reply:
Good catch! I changed the code to use memcpy then.